service/history/queue/queue_processor_util.go

```diff
@@ -0,0 +1,151 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package queue
+
+import (
+	"sort"
+	"time"
+
+	"github.com/uber/cadence/common/collection"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/service/dynamicconfig"
+	"github.com/uber/cadence/service/history/task"
+)
+
+type (
+	updateMaxReadLevelFn    func() task.Key
+	updateClusterAckLevelFn func(task.Key) error
+	queueShutdownFn         func() error
+
+	queueProcessorOptions struct {
+		BatchSize                           dynamicconfig.IntPropertyFn
+		MaxPollRPS                          dynamicconfig.IntPropertyFn
+		MaxPollInterval                     dynamicconfig.DurationPropertyFn
+		MaxPollIntervalJitterCoefficient    dynamicconfig.FloatPropertyFn
+		UpdateAckInterval                   dynamicconfig.DurationPropertyFn
+		UpdateAckIntervalJitterCoefficient  dynamicconfig.FloatPropertyFn
+		SplitQueueInterval                  dynamicconfig.DurationPropertyFn
+		SplitQueueIntervalJitterCoefficient dynamicconfig.FloatPropertyFn
+		QueueSplitPolicy                    ProcessingQueueSplitPolicy
+		RedispatchInterval                  dynamicconfig.DurationPropertyFn
+		RedispatchIntervalJitterCoefficient dynamicconfig.FloatPropertyFn
+		MaxRedispatchQueueSize              dynamicconfig.IntPropertyFn
+		MetricScope                         int
+	}
+)
+
+func newProcessingQueueCollections(
+	processingQueueStates []ProcessingQueueState,
+	logger log.Logger,
+	metricsClient metrics.Client,
+) []ProcessingQueueCollection {
+	processingQueuesMap := make(map[int][]ProcessingQueue) // level -> state
+	for _, queueState := range processingQueueStates {
+		processingQueuesMap[queueState.Level()] = append(processingQueuesMap[queueState.Level()], NewProcessingQueue(
+			queueState,
+			logger,
+			metricsClient,
+		))
+	}
+	processingQueueCollections := make([]ProcessingQueueCollection, 0, len(processingQueuesMap))
+	for level, queues := range processingQueuesMap {
+		processingQueueCollections = append(processingQueueCollections, NewProcessingQueueCollection(
+			level,
+			queues,
+		))
+	}
+	sort.Slice(processingQueueCollections, func(i, j int) bool {
+		return processingQueueCollections[i].Level() < processingQueueCollections[j].Level()
+	})
+
+	return processingQueueCollections
+}
+
+// RedispatchTasks should be un-exported after the queue processing logic
+// in history package is deprecated.
+func RedispatchTasks(
+	redispatchQueue collection.Queue,
+	taskProcessor task.Processor,
+	logger log.Logger,
```

Comment:
why it needs to pass in the logger

Reply:
Because I'd like to emit a log when re-dispatch failed and different queue processors are using different loggers (log tags are different).